accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,113 @@
+
+There's also a discussion to setup the build of runtime/libraries to automatically trim based on target configuration - [here](https://github.com/dotnet/runtime/issues/31785). The intent of this change is not feature switches, but it's in the same area.
+
+### Runtime/libraries already have compatibility switches
+For example `CoreLib` already has several switches like `Switch.System.Globalization.FormatJapaneseFirstYearAsNumber`. These are typically small "tweaks" to the behavior of the runtime/libraries. They are internally exposed as static properties on [`LocalAppContextSwitches`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs) with the value coming from `.runtimeconfig.json` via [`AppContext.TryGetSwitch`](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext.trygetswitch?view=netcore-3.1).
+
+
+## Proposal
+The purpose of the proposed changes is to introduce patterns and functionality to support them to make introduction of new feature switches easier and consistent.
+
+### Unified pattern for defining switches and their properties
+Each feature switch should have a read-only static property defined and all the decisions about the feature's availability should be based on a simple condition using such property. Depending on the impact of the feature switch the property may or may not be a public API.
+
+Feature switches for large areas of functionality which may span multiple libraries should be exposed as public APIs to make it easy and consistent. For example if there would be a feature switch to disable support for HTTP2 there should be a public API property indicating the desired behavior. This is because it's likely that even libraries outside of core libraries may need to change their behavior based on the property. Having one property in this case would enable linker to trim branches across multiple libraries in a consistent way.
+
+Feature switches for localized functionality which is not likely to span multiple libraries can use only internal properties.
+
+*Note: `CoreLib` for example already has `LocalAppContextSwitches` which does basically exactly this. We could simply extend it to usage of feature switches not just compatibility switches.*
+
+*Question: It's unclear why `GlobalizationMode.Invariant` is not based on this behavior - but maybe it needs native runtime integration.*
+
+### Unified pattern for feature definition in MSBuild
+Introduce a standard pattern how to define these properties in the SDK. Basically a way to define a property and have it automatically passed through to `.runtimeconfig.json` as well as linker substitutions.
+
+This requires a mapping between the MSBuild property name, the full name of the feature switch for runtime configuration (which would show up in `.runtimeconfig.json`) and the read-only static property in the managed code which is used to branch the behavior. 
```

Comment:
How do you expect this mapping to surface and be exposed to the linker?

Reply:
Depending on the picked representation (see below) either the linker would be the one knowing about the mapping (attributes or pattern matching) or it would be directly given the result of the mapping. Linker should not be the component which has the mapping hard-coded - it should be hard-coded either in the SDK, or in the libraries code or nowhere (linker pattern matching).

Reply:
I think we should be careful about what we decide to expose. Make it easy to tweak any static bool property might cause developers and us pain when they decide to play with fire and become framework optimizers. My suggestion would be to expose this in two ways

1. Hidden
We'll use the properties (substitutions) to simplify our implementation. In that case, we need to ship the rules together with the framework assemblies (probably embedded resources) and they cannot be overwritten (think of big-endian for win-x64) and should not be available for public modification.

2. Visible
We expose a set of predefined well tested features like (globalization or serialization) and only they will have any meaning to the framework. We map them all into same MSBuild property or something similar and the features will be quite high-level and validated by us that they work as expected alone or when combined together.

So in both cases, the mapping should be part of the assembly resources as it can be RID specific or in an internal attribute if we want to start inventing another attribute set and add another linker pass to read them.

Reply:
If we think we need to make this explicit (as in only specific set of properties can be treated as consts by the linker) then I would very much like to use attributes for it. The overall number of them will be tiny and linker doesn't need a new pass - the `TypeMapStep` should be the right place.
Note that if we keep the support for `substitution.xml` in the linker in theory developers can play these games anyway. So it's just a matter of how "hidden" it is.
I personally would go with one of these two:
* Use the name of the IL property as the name of the runtime property - and only have a mapping from MSBuild property to this value. We would document the limited set of MSBuild properties which do have a mapping like this. Developers would have to play with the underlying runtime property set to get into dangerous places.
* Use attributes to describe the mapping from runtime property -> IL property

Reply:
Based on some questions that came up in https://github.com/mono/linker/pull/1047, I think we need more clarity on what this looks like for hidden switches in particularly. My understanding is that public features have a documented MSBuild property and feature name == appcontext string. What's an example of something we might make a hidden feature? I think endianness falls more into the category of https://github.com/dotnet/runtime/issues/31785 which is not a goal of the feature switches.

Reply:
@marek-safar to make the discussion easier - could you please provide a sample (probably from the Xamarin world) of a hidden "feature switch" which is not tied to a build of libraries, but would be provided by the Xamarin SDK for example?

As for how we provide the mapping:
* I personally like attributes better as they are naturally tied to the code.
* But I do see value in discussing the embedded XML as that would avoid another pass in the linker (which is pretty expensive).

Reply:
> when they decide to play with fire and become framework optimizers

With great power comes great responsibility! 😎!

Could the concept of a feature switch be tied to API checks, essentially, they are availability checks to change behaviour at runtime and compile time? Is it not?

Reply:
@Nirmal4G I must admit I don't know what exactly you refer to by "API checks". The proposal effectively uses public static bool API properties as the tool which the app/library/framework can use to determine behavior and linker will use it to trim the code.

I do agree about power and responsibility, but if possible we should not make it easy to break things. So the ability to play god will still be there (via custom substitution.xml files), but it's not going to be as simple as defining a single MSBuild property.

Reply:
_Again, its philosophical. but does have a some context here._

We need that power when deploying .NET to resource constrained devices. e.g. IoT and Apple Watch, etc.

> if possible we should not make it easy to break things

True. I like having the file to change the linker behaviour. But even for simple ones I opt for the file way rather than MSBuild properties. MSBuild Items can be used instead too.

Reply:
> But I do see value in discussing the embedded XML as that would avoid another pass in the linker

What would be the new pass? My naive expectation is that the custom attribute would be processed when we do `RemoveUnreachableBlocks`: before evaluating the body of the method for its "constantness" in `FindConstantExpressionsMethods`, we would look for the attribute and treat it as a constant if it has one and the feature is being removed. Feature switch already depends on `RemoveUnreachableBlocks` running.